aiida-yambo blogpost (#92)
aiidateam/aiida-website: docs/news/posts/2025-05-14-aiida-yambo.md

---
blogpost: true
category: Blog
tags: plugins, success-stories
author: Miki Bonacci
date: 2025-05-14
---

# AiiDA success stories: band structures and excitons with aiida-yambo

> *An example on how AiiDA is used to enable science.*

Author: I agree (just to clarify, enable is just an emphasis on the fact that without such tools we could not do science like HT - MC3D and stuff like that)

> **Note:** For more details on the Yambo code features and usage, please consult the [official website](https://www.yambo-code.eu/) and the corresponding [Yambo wiki page](https://wiki.yambo-code.eu/wiki/index.php?title=Main_Page).

To give you an idea of the complexity of typical MBPT simulations, in Figure 1 we show an AiiDA provenance graph for the G0W0 quasiparticle convergences workflow of a single material.

![Figure 1: YamboConvergence workflow provenance graph.](../pics/GW_conv_graph.png)

*Figure 1: Provenance graph of a G0W0 quasiparticle convergence workflow for a single material, generated using the `aiida-yambo` plugin. This graph highlights the complexity and interdependence of the calculations involved in MBPT simulations.*

Starting from the input set (on the left of the figure), we then perform several interconnected calculations (DFT+GW) which are then used to estimate the convergence set of parameters which allows us to obtain accurate quasiparticle corrections (i.e. band structure)

## Case studies

Before showing the plugin and a taste of its usage, we show here some of the applications which were made possible thanks to the `aiida-yambo` plugin.

> One of the first studies performed within the plugin was the calculation of excitonic effects in the 2D C3N ([Bonacci et al., Phys. Rev. Materials, 6:034009, 2022](https://journals.aps.org/prmaterials/abstract/10.1103/PhysRevMaterials.6.034009)), a material particularly promising for optoelectronic application.

Reviewer: Hmmm, this URL:
https://journals.aps.org/prmaterials/abstract/10.1103/PhysRevMaterials.6.034009

breaks in linkcheck with:

```
/home/runner/work/aiida-website/aiida-website/docs/news/posts/2025-05-14-aiida-yambo.md:48: WARNING: broken link: https://journals.aps.org/prmaterials/abstract/10.1103/PhysRevMaterials.6.034009 (403 Client Error: Forbidden for url: https://journals.aps.org/prmaterials/abstract/10.1103/PhysRevMaterials.6.034009)
```

Maybe there's an arxiv link?

Author: yeah but I would like to cite the PRM

Author: I put the DOI

Reviewer: Good idea!

The `aiida-yambo` plugin has been instrumental in several other studies:

- [Marrazzo et al., "Relative Abundance of Z2 Topological Order in Exfoliable Two-Dimensional Insulators", Nano Letters, 19:8431-8440 (2019)](https://pubs.acs.org/doi/abs/10.1021/acs.nanolett.9b02689);
- [Atambo et al., "Electronic and optical properties of doped TiO2 by many-body perturbation theory", Phys. Rev. Materials 3, 045401 (2019)](https://doi.org/10.1103/PhysRevMaterials.3.045401);
- [Betti et al., “Gap opening in double-sided highly hydrogenated free-standing graphene", Nano Letters, 22(7):2971–2977 (2022)](https://pubs.acs.org/doi/full/10.1021/acs.nanolett.2c00162);
- [Betti et al., “Dielectric response and excitations of free-standing graphane”, Carbon Trends 12, 100274 (2023)](https://doi.org/10.1016/j.cartre.2023.100274);
- [Bonacci et al., "Benchmarking the GW100 dataset with the Yambo code by means of G0W0 approximation, Materials Cloud Archive 2023.54 (2023)"](https://doi.org/10.24435/materialscloud:ce-82).

## How to Run a simple GW Calculation with `aiida-yambo`

> The `YamboWorkflow` is the core workchain of the `aiida-yambo` plugin. Here below, we consider the case of a single GW calculation.

Reviewer: ```suggestion
The `YamboWorkflow` is the core workchain of the `aiida-yambo` plugin. Below, we consider the case of a single GW calculation.
```

> > **Note:** Running automated BSE@GW and several other type of simulations, e.g. convergences, are also supported and are covered in the dedicated [set of tutorials](https://aiida-yambo.readthedocs.io/en/documentation/user_guide/index.html).

Reviewer: ```suggestion
> **Note:** Running automated BSE@GW and several other types of simulations, e.g. convergence studies, are also supported and are covered in the dedicated [set of tutorials](https://aiida-yambo.readthedocs.io/en/documentation/user_guide/index.html).
```

> The workflow orchestrates all steps involved in a typical GW calculation: from initial DFT preprocessing to the final GW calculation, ensuring smooth integration between **Quantum ESPRESSO** and **Yambo** codes.

Reviewer: ```suggestion
The workflow orchestrates all steps involved in a typical GW calculation: from initial DFT preprocessing to the final GW calculation, ensuring smooth integration between the **Quantum ESPRESSO** and **Yambo** codes.
```

Depending on the input provided, the workflow will:

> - Launch SCF and NSCF DFT calculations using `PwBaseWorkChain` (from the `aiida-quantumespresso` plugin)

Reviewer: ```suggestion
- Launch SCF and NSCF DFT calculations using the `PwBaseWorkChain` (from the `aiida-quantumespresso` plugin)
```

- Run the GW step using `YamboRestart` (automatic error handling)
- Skip already-computed steps if valid parent input is provided, leveraging data provenance

The execution is dynamic: users don’t need to explicitly define each intermediate step unless they want to.

> If any underlying process fails (e.g., due to memory errors), the workflow exits cleanly and transparently thanks to the robust restart mechanisms inherited from `BaseRestartWorkChain`.

Author: no, there is always a restart mechanism, but it can fail of course... I added "and cannot be restarted successfully"

The only required inputs are:

- the codes we are going to use during the simulation: pw.x, p2y and yambo (as AiiDA `Code` nodes);
- the input structure (as AiiDA `StructureData` node).

This snippet of code is what you need to perform a basic GW calculation within the plugin:

```python
from aiida import orm, load_profile
load_profile()

from aiida.plugins import WorkflowFactory
from aiida.engine import submit

YamboWorkflow = WorkflowFactory('yambo.yambo.yambowf')

# ...
# Loading the codes and the structure here
# ...

> builder = YamboWorkflow.get_builder_from_protocol(

Author: default is "moderate" (still old namings... I put protocol="fast" here for simplicity)

    pw_code = pw_code,
    preprocessing_code = p2y_code,
    code = yambo_code,
    structure= structuredata_node,
)

run = submit(builder)
```

it is possible to ask the workflow to parse specific quantities, like gaps or single quasiparticle levels.
To do that, we can provide an additional input to our builder. For example, to ask for the minimum gap, the direct gap at Gamma and the HOMO and LUMO:

```python

> builder.additional_parsing = List(list=['gap_','gap_GG','homo','lumo'])

Reviewer: ```suggestion
builder.additional_parsing = orm.List(['gap_','gap_GG','homo','lumo'])
```
Should be possible, no?

Author: yes, added

```

At the end of the calculation we obtain an output node, called `output_ywfl_parameters`, containing the information we asked the workflow to compute (as well as their DFT counterpart):

```python

> {'gap_': 1.0573099312007,
>  'homo': -0.30465849325061,
>  'lumo': 0.75265143795013,
>  'gap_GG': 3.0613290492713,
>  'homo_G': -0.30465849325061,
>  'lumo_G': 2.7566705560207,
>  'gap_dft': 0.66473089347482,
>  'homo_dft': 0.0,
>  'lumo_dft': 0.66473089347482,
>  'gap_GG_dft': 2.5552411541998,
>  'homo_G_dft': 0.0,
>  'lumo_G_dft': 2.5552411541998}

Reviewer: ```suggestion
{
    "gap_": 1.0573099312007,
    "homo": -0.30465849325061,
    "lumo": 0.75265143795013,
    "gap_GG": 3.0613290492713,
    "homo_G": -0.30465849325061,
    "lumo_G": 2.7566705560207,
    "gap_dft": 0.66473089347482,
    "homo_dft": 0.0,
    "lumo_dft": 0.66473089347482,
    "gap_GG_dft": 2.5552411541998,
    "homo_G_dft": 0.0,
    "lumo_G_dft": 2.5552411541998,
}
```

```

## Conclusions

> The integration of Yambo with AiiDA through the aiida-yambo plugin has significantly streamlined MBPT simulations, making them more accessible and reproducible.

Reviewer: ```suggestion
The integration of Yambo with AiiDA through the aiida-yambo plugin provides highly streamlined MBPT simulations, making them more accessible and reproducible.
```
